### DISCUSSION TITLE
Implement BGP

### DISCUSSION THREAD

**Discussion by sebiklamar** [vehagn/homelab#156] [Ideas]
Hi Vegard,

as I saw you having BGP support on your roadmap list, I want to share my documentation and experience of my BGP w/ Opnsense including `EgressGateway` PoC.

# What's in the pocket

1. **IN:** `Ingress` (`IngressGateway`) with any IP you want, i.e. you don't need to have a matching interface for that. Just name an IP to your liking in the `io.cilium/lb-ipam-ips` annotation as usually and cilium will add it to the BGP routing.
2. **IN:** `L2` routing/reachability of your BGP-managed IPs because your default gateway will do a RA for your hosts still residing in the same L2 net (your `192.168.1.0/24` subnet), i.e. you can switch of your `L2Announcement` code. I didn't further investigate this b/c this is not a relevant scenario in my environment (k8s nodes in a separate server VLAN).
3. **IN w/ _workaround_:** `Egress` (`EgressGateway`) with BGP iff you have an `Ingress` defined in parallel. Due to the limitation mentioned in the next (4.) item there's no BGP advertisement for the `Egress`. Hence, you need to define an `Ingress` in parallel (counter-measure 1) and you need to ensure the `Egress` is set up for the same node as the pod using the `Ingress` is running (counter-measure 2) , i.e. you will need to pin (`nodeSelector`) both the `Deployment` and the `CiliumEgressGatewayPolicy`.
   NB: For `Ingress`-only scenario (1.) no pinning is needed at all, i.e. you will have full flexibility.
4. **OUT:** `Egress`-only scenario (without any `Ingress` for the same IP). Would need Cilium Enterprise Ed., cf. [Cilium blog post for Cilium Enterprise 1.15](https://isovalent.com/blog/post/isovalent-enterprise-for-cilium-1-15/#h-bgp-support-for-egress-gateway).
5. **INCOMPATIBLE:** If you want to use `EgressGateway` feature, you would need to live without `CiliumEndpointSlice` feature (cf. [cilium issue 24833](https://github.com/cilium/cilium/issues/24833)). `CiliumEndpointSlice` is currently enabled in your environment.

# How-to

## TLDR

For seeing all pieces in action see in my evolving [homelab repo](https://github.com/sebiklamar/homelab/) based on your setup

- [`openssh`](https://github.com/sebiklamar/homelab/tree/main/k8s/apps/ras/openssh/envs/dev) app for `EgressGateway` (and `IngressGateway`) example.
- [`additional-values.yaml`](https://github.com/sebiklamar/homelab/blob/main/k8s/core/network/cilium/envs/dev/additional-values.yaml) file for cilium installation, cf. details below.
- the cilium [`base`](https://github.com/sebiklamar/homelab/blob/main/k8s/core/network/cilium/base) and [`envs/dev`](https://github.com/sebiklamar/homelab/blob/main/k8s/core/network/cilium/envs/dev/) folders contain the Resource definitions for BGP

## Links

- Opnsense BGP: https://docs.opnsense.org/manual/dynamic_routing.html#bgp-section
- Cilium BGP: https://docs.cilium.io/en/stable/network/bgp-control-plane/bgp-control-plane-v2/#bgp-control-plane-v2
- outdated (using `BGPPeeringPolicy`), though good opnsense docu: https://baremetalblog.com/posts/tech/2024-03-12-cilium-bgp-and-you/
- `EgressGateway`: https://docs.cilium.io/en/stable/network/egress-gateway/egress-gateway/

## opnsense guide

See the 3rd link listed above which is having good screenshot documentation on opnsense setup.

My documented install notes from Obsidian:

1. Install `os-frr` package (from _Plugins_ tab)
2. Open firewall port `tcp/179` on interface `K8S`
3. Configure Peers [[#Configuration]]

### Warning message during install

```
Beware that remote control of frr8 daemons over TCP sockets is enabled by
default.
Use daemon flags in /etc/rc.conf to disable it if unneeded, for example:
zebra_flags="-P0"
ospfd_flags="-P0"

FRR's OSPF daemons tries to allocate big socket buffer, so generate warning
messages like:
"setsockopt_so_sendbuf: fd 6: SO_SNDBUF set to 1048576 (requested 8388608)"
To prevent such message kern.ipc.maxsockbuf can be increased:
sysctl kern.ipc.maxsockbuf=16777216
```

#### Routing > General

| **Field**                  | **Value**              |
| -------------------------- | ---------------------- |
| Enabled                    | true                   |
| Enable CARP Failover       | true                   |
| Enable SNMP AgentX Support | default: false         |
| Enable Logging             | default: true          |
| Log Level                  | default: Notifications |
| Firewall rules             | default: true          |

##### General

| **Field**             | **Value**                                           |
| --------------------- | --------------------------------------------------- |
| Enabled               | true                                                |
| BGP AS Number         | 64520                                               |
| Network               | <blank>                                             |
| Log Neighbour Changes | default: false                                      |
| Route Redistribution  | Connected routes (directly attached subnet or host) |

##### Neigbours

| **Field**               | **Value**         |
| ----------------------- | ----------------- |
| Enabled                 | true              |
| Description             | dev-work-01 / -02 |
| Peer-IP                 | 10.7.8.134 / .135 |
| Remote AD               | 64523             |
| Update Source Interface | K8S               |
| Next-Hop-Self           | Enabled           |
| _all other fields_      | <blank>           |

## k8s guide

Ensure BGP is enabled in cilium

```yaml
# additional-values.yaml
bgpControlPlane:
  enabled: true
```

For also having `Egress` feature available, cilium `values.yaml` file needs to be at least:

```yaml
# additional-values.yaml
## enable BGP
# URL: https://docs.cilium.io/en/stable/network/bgp-control-plane/bgp-control-plane/#installation
bgpControlPlane:
  enabled: true

## enable egress gateway
# URL: https://docs.cilium.io/en/stable/network/egress-gateway/egress-gateway/
egressGateway:
  enabled: true
bpf:
  masquerade: true

# egress gateway not compatible with CiliumEndpointSlice
# URL: https://github.com/cilium/cilium/issues/24833
enableCiliumEndpointSlice: false
```

FYI: I currently maintain a _base_ `values.yaml` and an env.-specific `additional-values.yaml` file. Though, I didn't manage yet to maintain the `values.yaml` file only _once_ in the `base` folder with only _patches_ being applied in the environments as overlay.

#### Overview

- `base/CiliumBGPAdvertisements`: Defines prefixes that are injected into the BGP routing table.
- `base/CiliumBGPPeerConfig`: A common set of BGP peering setting. It can be used across multiple peers.
- `<env>/CiliumBGPClusterConfig`: Defines BGP instances and peer configurations that are applied to multiple nodes, specific to an `<environment>`.
- `CiliumBGPNodeConfigOverride`: Defines node-specific BGP configuration to provide a finer control, not used.

#### `<env>/CiliumBGPClusterConfig`

environment-specific

```yaml
---
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPClusterConfig
metadata:
  name: cilium-bgp
spec:
  nodeSelector:
    # disable BGP on control plane nodes
    matchExpressions:
      - key: node-role.kubernetes.io/control-plane
        operator: DoesNotExist

  bgpInstances:
    - name: "instance-64523"
      localASN: 64523 # Use your cluster's ASN here!
      peers:
        - name: "peer-64520-gw2"
          peerASN: 64520
          peerAddress: 10.7.8.2
          peerConfigRef:
            name: "cilium-peer"
        - name: "peer-64520-gw3"
          peerASN: 64520
          peerAddress: 10.7.8.3
          peerConfigRef:
            name: "cilium-peer"
```

## ASNs

Just pick an ASN ID from the private AS number range 64.512 – 65.534 (16b) or from the newer 32b range 4.200.000.000 - 4.294.967.294.

Most documented setups (incl. mine at the moment) use different ASNs for the router (e.g. opnsense) and partner (k8s cluster), although, AFAIK, the same ASNs can be used (it's the same party/system "Vegard's system/environment").

## Cheat Sheet

```sh
cilium bgp peers
cilium bgp routes
```

Have fun with a modern routing setup with certainly more robust connectivity (instead of the `k8sClientRateLimit` hack for L2), and true source IP.

HTH -- Sebastian

**Edit:** Added cilium folders links in TLDR section as openssh only covers Egress and not BGP and links.

**Comment by vehagn** [comment]
@sebiklamar Thanks for the info! This is a goldmine once I start testing it out.

I've been way too busy these last months, but I think I'll be able to look at BGP with Cilium and UniFi soon™

Other resources I have for this are
- https://baremetalblog.com/posts/tech/2024-03-12-cilium-bgp-and-you/
- https://www.packetswitch.co.uk/bgp/
- https://medium.com/@spudstr/ubiquity-unifi-k3s-bgp-and-metallb-744b50706c7c
- https://docs.cilium.io/en/latest/network/bgp-control-plane/bgp-control-plane-v2/

**Reply by ilbarone87** [reply]
Noticed you have implemented bgp routing in your cluster, are you planning to write an article about it? Also is there any benefit over l2announcement?

**Reply by vehagn** [reply]
@ilbarone87 I've started writing an article over at [GitLab](https://gitlab.com/vehagn/blog/-/tree/bgp/content/articles/2025/10/bgp-cilium-unifi). I'm hoping to finish it this week.
I haven't done any benchmarking, though there's some anecdotal evidence over at [baremetalblog](https://baremetalblog.com/posts/tech/2024-03-12-cilium-bgp-and-you/) that it reduces gaming lag.
Other potential benefits are less network overhead (no ARP broacasts) and better routes.

I don't think there's any real reason to do it in my use-case other than "just because" and nerd cred 😎

**Reply by ilbarone87** [reply]
Nice, I'll wait for your write up then as I have RKE2 with cilium and UniFi. Major issue I have with l2announcement is that doesn't support externalTrafficPolicy:Local  https://github.com/cilium/cilium/issues/27800 and affects the visibility of source IPs in requests (running Pi-hole in the cluster and all I can see are 10.42.x.x IPs, k8s ip range)

**Comment by sebiklamar** [comment]
Yeah, @ilbarone87 , leveraging `externalTrafficPolicy: Local` for preserving source IP was also one of the big motivators for me for BGP.
There are also downsides, for sure. I've encountered a minor issue in terms of an interruption of connectivity for the edge case when having more than one router.  I say minor because that case should not be that common in homelabs – though in enterprise environments.

Background: I'm running 2 OPNsense instances as core routers in HA mode for not only having a test/backup instance but also having a fail-over when the primary router is rebooting for keeping connectivity across VLANs and upstream.

Issue: BGP routes take a while (15 - 30 s) to establish on the 2nd router (when it becomes primary) as the K8S cluster is only advertising them to primary router peer.  So, while I currently do not lose even a single PING packet in my network, I currently have an interruption of connectivity to services hosted in K8S (which are only 2 at the moment).

Outlook: Possible options could be decreasing the transition time or maybe K8S advertising its routes to all peers at the same time.

FYI: Will likely put it on my ToDo list as I put DNS on K8S recently.  There's a secondary DNS resolver available anyway, however it puts some delay on the processing if the K8S one is used as first resolver and when there's no parallel processing from the client (which is default in most OS).  (Right now focussing still on proxmox-csi backup methods and monitoring of cluster for having control on PVC disk usage, i.e. the foundations of the cluster)

Other than that... I did not use L2 so far, so no real comparison. What maybe also could be an "inconvenience" for some is not having ICMP PING available for services running with a BGP-advertised LB IP (cf. https://github.com/cilium/cilium/issues/14118).

**Comment by vehagn** [comment]
It took way too much time, but I think I've finished! https://blog.stonegarden.dev/articles/2025/11/bgp-cilium-unifi/

**Reply by ilbarone87** [reply]
Noticed you had issue with l2 https://github.com/vehagn/homelab/commit/3b812ed476b06dcb84939679159abe374c4f4bd5 what was the problem?

**Reply by vehagn** [reply]
I'm experiencing some weird packet drops from my ISP.

The pages exposed directly works perfectly on my work network, but when trying to connect through a mobile carrier I experience that a lot of packets are dropped from at least two different carriers. I've also had problems from other people's networks from different ISPs.

The pages being served through a Cloudflare tunnel experiences no issues what so ever.

I wanted to rule out BGP being an issue, so I reverted to ARP L2 announcements for a while. I just haven't cleaned it up yet.
